pulser-core/pulser/operators/operator.py

```diff
@@ -0,0 +1,419 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classes to store Operators for simulation and measurements."""
+
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections import UserList
+from collections.abc import Collection
+import itertools
+from dataclasses import dataclass
+from numbers import Number
+from typing import Mapping
+
+from pulser.channels import State, STATES_RANK
+from pulser.register import QubitId
+
+PROJECTORS = {"sigma_" + a + b for a in STATES_RANK for b in STATES_RANK}
+
+
+def _check_kron_object(
+    kron_obj: tuple[tuple[QuditString | QuditOperatorString, set[QubitId]]],
+    obj_type: type,
+) -> None:
+    """Check if an object with Kronecker products is correctly defined."""
+    if not isinstance(kron_obj, tuple) or any(
+        [isinstance(operation[0], obj_type) for operation in kron_obj]
+    ):
+        raise ValueError(
+            "The kron_"
+            f"{"qudits" if str(obj_type) == "QuditString" else "operations"}"
+            " should be a tuple of tuples of shape (operator as a "
+            f"{obj_type}, qubit id as an int or str)."
+        )
+    # Check the qubit ids targeted in operations
+    target_qids = [operation[1] for operation in kron_obj]
+    if "global" in target_qids:
+        if len(kron_obj) > 1:
+            raise ValueError(
+                "If a 'global' operation is defined, no other operations "
+                "can be defined."
+            )
+    else:
+        target_qids_list = list(itertools.chain(*target_qids))
+        target_qids_set = set(target_qids_list)
+        if len(target_qids_set) < len(target_qids_list):
+            # Either the qubit id has been defined twice in an operation:
+            for qids in target_qids:
+                if len(set(qids)) < len(qids):
+                    raise ValueError("Duplicate atom ids in argument list.")
+            # Or it was defined in two different operations
+            raise ValueError(
+                "Each qubit can be targeted by only one operation."
+            )
+
+
+@classmethod
+class QuditString:
+    """A 1-qudit state."""
+
+    coefficients: list[Number]
+    states: list[State]
+
+    def __post_init__(self):
+        for state in self.states:
+            if state not in STATES_RANK:
+                raise ValueError(f"State {state} does not exist.")
+
+    def as_tuple(self) -> tuple[tuple[Number, State]]:
+        """Returns tuple of (coefficient, state label)."""
+        return zip(self.coefficients, self.states)
```

Comment:
`tuple(zip(...))`? `zip` or change the type hint since `zip` returns an iterator